### PR TITLE
Catch errors thrown in ensure block

### LINKED ISSUE #78: Throwing an error in an `ensure()` hook can leave effection tree in an unstable state
It's a bug to have an error in a teardown callback, but effection currently does not catch this, which ends up blowing up the entire teardown process leaving the tree in an inconsistent state.

We should catch errors like this and write a critical warning to the console.

### PR DESCRIPTION
**Pull request by jnicklas**
Closes #78

If an error ever occurrs in an ensure block, then that is a serious problem. We should inform the user that this is not good by printing a huge nasty warning to the console.

However, if such a situation does occur, at least it should not leave the process tree in an inconsistent state. Other exit hooks should still run, and other processes should still shut down. Currently if such an error does occur, it will blow the stack, and cause finalization to stop
abrubtly. This is very much not good.

Here we're rescuing any such errors, printing them to the console, but crucially we are NOT re-raising them. This allows us to proceed with finalization, which will leave the process tree in a more consistent state.

It's worth noting though that such a situation really is very undesirable. The Rust language does a hard abort of the process with an error exit code if a destructor panics, we cannot do something similar in effection, since we want to run in browsers, and browsers cannot hard-exit.

### CODE REVIEW

**Comment by frontsidejack** [comment]
A preview package of this pull request has been released to NPM with the tag `catch-ensure-errors`.
You can try it out by running the following command:
```bash
$ npm install effection@catch-ensure-errors
```
or by updating your package.json to:
```bash
{
  "effection": "catch-ensure-errors"
}
```
Once the branch associated with this tag is deleted (usually once the PR is merged or closed), it will no longer be available. However, it currently references [effection@0.5.1-b0b0b5f](https://www.npmjs.com/package/effection/v/0.5.1-b0b0b5f) which will be available to install forever.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 6ca7c79b292540d811f4cfacd2af83a6a8be09f3
</p>

**Comment by frontsidejack** [comment]
A preview package of this pull request has been released to NPM with the tag `catch-ensure-errors`.
You can try it out by running the following command:
```bash
$ npm install effection@catch-ensure-errors
```
or by updating your package.json to:
```bash
{
  "effection": "catch-ensure-errors"
}
```
Once the branch associated with this tag is deleted (usually once the PR is merged or closed), it will no longer be available. However, it currently references [effection@0.5.1-474a0b2](https://www.npmjs.com/package/effection/v/0.5.1-474a0b2) which will be available to install forever.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 9dc53336b59443bcf7a9e41540c9bc5d27aca461
</p>

**Comment by frontsidejack** [comment]
A preview package of this pull request has been released to NPM with the tag `catch-ensure-errors`.
You can try it out by running the following command:
```bash
$ npm install effection@catch-ensure-errors
```
or by updating your package.json to:
```bash
{
  "effection": "catch-ensure-errors"
}
```
Once the branch associated with this tag is deleted (usually once the PR is merged or closed), it will no longer be available. However, it currently references [effection@0.5.1-10bb04b](https://www.npmjs.com/package/effection/v/0.5.1-10bb04b) which will be available to install forever.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 5867178c61724fff8dcc02ebd439058f675db90f
</p>

**Comment by frontsidejack** [comment]
A preview package of this pull request has been released to NPM with the tag `catch-ensure-errors`.
You can try it out by running the following command:
```bash
$ npm install effection@catch-ensure-errors
```
or by updating your package.json to:
```bash
{
  "effection": "catch-ensure-errors"
}
```
Once the branch associated with this tag is deleted (usually once the PR is merged or closed), it will no longer be available. However, it currently references [effection@0.5.1-312a6d5](https://www.npmjs.com/package/effection/v/0.5.1-312a6d5) which will be available to install forever.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 8ec127f427554f9d063064e4817eec50abba295f
</p>

**Comment by jnicklas** [comment]
@cowboyd not sure why the type tests are failing. It looks unrelated to this PR though.

File: src/context.js [hunk 1]
```diff
@@ -135,7 +135,17 @@ Thanks!`);
...
     for (let hook of [...this.exitHooks].reverse()) {
-      hook();
+      try {
+        hook();
+      } catch(e) {
+        /* eslint-disable no-console */
+        console.error(`
+CRITICAL ERROR: an exception was thrown in an exit handler, this might put
+Effection into an unknown state, and you should avoid this ever happening.
+Original error:`);
```
Comment (cowboyd) [inline src/context.js]:
```suggestion
        console.error(`
CRITICAL ERROR: an exception was thrown in an ensure() callback. This
could mean that resources are not properly deallocated and that the
system is left in an unknown state. ensure() callbacks should never
raise exceptions, so this is probably an application level bug.
Original error:`);
```

**Comment by cowboyd** [comment]
> @cowboyd not sure why the type tests are failing. It looks unrelated to this PR though.

I'm guessing that it's because we have our TS config pointed to `esnext` which is a moving target..... which moved. Maybe if we peg it to `es2019` or some fixed target?

**Comment by frontsidejack** [comment]
A preview package of this pull request has been released to NPM with the tag `catch-ensure-errors`.
You can try it out by running the following command:
```bash
$ npm install effection@catch-ensure-errors
```
or by updating your package.json to:
```bash
{
  "effection": "catch-ensure-errors"
}
```
Once the branch associated with this tag is deleted (usually once the PR is merged or closed), it will no longer be available. However, it currently references [effection@0.5.1-2c504ef](https://www.npmjs.com/package/effection/v/0.5.1-2c504ef) which will be available to install forever.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 1ec6b7a10226e7ceb5e6059653e3ad6c724c6954
</p>

**Comment by frontsidejack** [comment]
A preview package of this pull request has been released to NPM with the tag `catch-ensure-errors`.
You can try it out by running the following command:
```bash
$ npm install effection@catch-ensure-errors
```
or by updating your package.json to:
```bash
{
  "effection": "catch-ensure-errors"
}
```
Once the branch associated with this tag is deleted (usually once the PR is merged or closed), it will no longer be available. However, it currently references [effection@0.5.1-5826040](https://www.npmjs.com/package/effection/v/0.5.1-5826040) which will be available to install forever.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 1ab9ce5bab504af111e6284d132934bb3ba9578d
</p>

**Comment by jnicklas** [comment]
:shipit:

**Pull request merged**